### ISSUE TITLE
Standard Models lab fails 

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#2278]
Following successful configuration, start (and verified both cts and asset management notebooks work ok)

Standard models lab fails as reported at https://odpi.slack.com/archives/CBPRDQPRC/p1576732596001800

 - this report was with docker compose
 - also fails on localhost

<img width="1064" alt="Screenshot 2019-12-19 at 12 42 37" src="https://user-images.githubusercontent.com/7292002/71174457-25d15480-225d-11ea-87ae-d99f41a5b127.png">

**Comment by planetf1** [comment]
Response from http request is `{'relatedHTTPCode': 200, 'result': []}`

**Comment by mandy-chessell** [comment]
There are acrually 2 HTTP requests being issued here and it is the 2nd one that is failing.  If I change the lab notebook to print out the results of each HTTP request as follows:

```
archiveFileName = 'CloudInformationModel.json'
loadArchiveURL = serverPlatformURL +  '/open-metadata/admin-services/users/' +  adminUserId + '/servers/' + server + '/instance/open-metadata-archives/file'

print(" ")
response = requests.post(loadArchiveURL, archiveFileName)
print ("Returns:")
prettyResponse = json.dumps(response.json(), indent=4)
print (prettyResponse)
print (" ")

getGlossaryURL = serverPlatformURL + '/servers/' + server + '/open-metadata/access-services/glossary-view/users/' +  erinsUserId + '/glossaries'

response = requests.get(getGlossaryURL)

print ("Returns:")
prettyResponse = json.dumps(response.json(), indent=4)
print (prettyResponse)
print (" ")


result = response.json().get('result')
glossary = result[0]
glossaryId = glossary.get('guid')
print ("GlossaryId: " + glossaryId)
print (" ")
glossaryUsage = glossary.get('usage')
glossaryDisplayName = glossary.get('displayName')
print (glossaryDisplayName + ": " + glossaryUsage)
glossaryDescription = glossary.get('description')
print (" ")
print (glossaryDescription)
```
I see 
```
Returns:
{
    "relatedHTTPCode": 200
}
 
Returns:
{
    "relatedHTTPCode": 500,
    "exceptionClassName": "GlossaryService",
    "exceptionErrorMessage": "Unable to retrieve repository helper",
    "exceptionSystemAction": "getTypeDefGUID",
    "result": [],
    "actionDescription": "getTypeDefGUID"
}
```
So the notebook should be more resilient to:
* the fact that in the archive will not load if it has not been created.
* if the archive does not load, the list of glossaries returned is an empty list
* regressions in the Egeria function.

The failing call is a regression in the Glossary View OMAS.   The getAllGlossaries call is not working and the resulting error is incorrectly filled out:
* Exception class name should be the name of the exception class for the client to recreate
* Exception error message should include a unique Id to pinpoint the failure to a point in the code
* Exception system action should describe what the system did
* Exception user response is missing.

**Comment by mandy-chessell** [comment]
I have fixed the notebook to handle bad responses from the metadata server.
Also fixed the Glossary View OMAS to return meaningful error messages.  I now see:

```
Response was:
{
    "relatedHTTPCode": 404,
    "exceptionClassName": "org.odpi.openmetadata.commonservices.multitenant.OMAGServerInstance",
    "exceptionErrorMessage": "OMAG-PLATFORM-404-002 The Glossary View service is not available on OMAG Server cocoMDS2 to handle a request from user erinoverview",
    "exceptionSystemAction": "The system is unable to process the request.",
    "exceptionUserAction": "Verify that the correct server is being called on the correct platform and that the requested service is configured to run there.  Once the correct environment is in place, retry the request.",
    "result": [],
    "actionDescription": "getTypeDefGUID"
}
```
This is the root cause of the 500 error I documented earlier - which is actually a problem only in my build where I have been fixing up the names of the access services so that they each use a consistent name - missed one in the Glossary View OMAS :)

**Comment by planetf1** [comment]
I will retest once update merged

**Comment by mandy-chessell** [comment]
Thanks Nigel :) We also need to find a simple way to make the archive available

**Comment by planetf1** [comment]
> Thanks Nigel :) We also need to find a simple way to make the archive available
From the notebook it appears the archive file is being loaded by the server itself

If this is the case, of course we could place the file in a known good location - at least for our container environments.

However...

If that is the case how would the archive normally be deployed? I can understand for the CIM we may provide, but in other cases it's likely distributed independently from egeria itself?

As such I wonder if it's more appropriate to pass the contents of the file via the client - ie over rest? 

In that case we would want to make sure we allow the client (ie the notebook server in this case) to have access to the file?

**Comment by planetf1** [comment]
One tactical option for this notebook if we need to address in the 1.3 timeframe would be to direct the reader to use the 'upload' option within the notebook UI to upload a datafile to the current working directory (top level of notebooks) and then refer to that in the notebook.

Not ideal but would at least allow someone to work, then we can move the proper fix to ensure the file is shared in all contexts into 1.4 ?

**Comment by mandy-chessell** [comment]
@planetf1 you are right - the archive is loaded by the server.  It can be loaded at server start up, or by a REST call, but the archive must be accessible to the server.

Building the archive is done by the `CloudInformationModelArchiveWriter`.  The user of the lab needs to have downloaded the CIM model and then runs the archive write, then they can run the lab.  We could change our lab environment to have the archive ready built so the lab works - or work out a way to drive the whole process from the lab.

I have put a message in the lab to remind user to build the archive first.

**Comment by planetf1** [comment]
Ok. I suggest we'll go with the doc updates for 1.3, and get the files in place for 1.4. When we've shipped 1.3 I'll walk through the lab and get the resources in place.

**Comment by planetf1** [comment]
As per recent discussion on status call, we should add a stronger marker to this notebook to indicate it isn't completed yet - just to avoid user confusion. 

A specific suggestion was some large text/note right at the top of the notebook - under the logo.

There is a note saying the archive needs to be imported, but we've not yet really finished everything to make that easy in the notebook environment, so easiest to just warn people off - unless they want to get deeply involved in this area.


This may apply to other notebooks

Worth doing for 1.3 to avoid user confusion

**Comment by planetf1** [comment]
Taking a look at this:
 * CloudInformationModel.json is not currently in the egeria-omag-1.6-SNAPSHOT

**Comment by planetf1** [comment]
Also noted that the build process does not currently create CloudInformationModel.json !
It is referenced at https://github.com/odpi/egeria/blob/f552a7bb6abc75d24cac8d2892dffed977da7f98/open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/cloudinformationmodel/CloudInformationModelArchiveWriter.java

Will look at modifying the pom in https://github.com/odpi/egeria/blob/f552a7bb6abc75d24cac8d2892dffed977da7f98/open-metadata-resources/open-metadata-archives/design-model-archives/pom.xml to build this archive by default

**Comment by planetf1** [comment]
Revisiting above comments -- the instructions may refer to building the archive first, but in a container environment they would need to then get it deployed to the server.

I think we should take one of two approaches
 * pre-build (as above) - and distribute -- this still doesn't prevent a user from building themselves -- and makes it easier to get more content into egeria
 * Allow for the definitions to be uploaded to the server (body) - though this opens up security challenges potentially.

I think the former is most appropriate - will take a look cc: @mandy-chessell

**Comment by planetf1** [comment]
A few challenges:

* The design model tooling does not have a manifest with an entry point defined - so it cannot so easily be run via 'java -jar'.
* The jar actually contains multiple tools
```
➜  design-model-archives git:(issue2278) ✗ java -jar target/design-model-archives-1.6-SNAPSHOT.jar org.odpi.openmetadata.archiveutilities.designmodels.cloudinformationmodel.CloudInformationModelArchiveWriter
no main manifest attribute, in target/design-model-archives-1.6-SNAPSHOT.jar
```
* The classpath for egeria tooling can be complex - we have a lot of dependencies. However this jar does not include required dependencies, making it tricky to execute on the command line even if the entry point is overridden:
```
➜  design-model-archives git:(issue2278) ✗ java -cp target/design-model-archives-1.6-SNAPSHOT.jar org.odpi.openmetadata.archiveutilities.designmodels.cloudinformationmodel.CloudInformationModelArchiveWriter
Error: Could not find or load main class org.odpi.openmetadata.archiveutilities.designmodels.cloudinformationmodel.CloudInformationModelArchiveWriter
Caused by: java.lang.NoClassDefFoundError: org.odpi.openmetadata.repositoryservices.archiveutilities.OMRSArchiveWriter
```

A few things we could do here
* Full specify the classpath - near impossible at the CLI - though easy in IDE. Also possible to do within a maven build target fairly easy
* Include dependencies required by the tool -- this seems essential, and would at least make it possible to run the require dutility
* Optionally split up into multiple jars, with appropriate manifests so that each tool is standalone - easier to use

I propose to do the second for this change -- otherwise useage is too complex.
We should consider the third - maybe a future change

**Comment by planetf1** [comment]
Made changes and testing ...
 - not (yet) made utilities executable by adding dependencies - but they are invoked in build
 - not updated any docs, including notebook

**Comment by planetf1** [comment]
Tested

* Glossary loads ok:
```
Returns:
{
    "class": "VoidResponse",
    "relatedHTTPCode": 200
}
 
GlossaryId: af1a65f2-7b7c-4de7-8174-944e236ee38a; Cloud Information Model (CIM): Data types for commerce focused cloud applications.
 -> The Cloud Information Model (CIM) is an open source glossary and data model that spans the following subject areas:

•	Party – people, their roles and organizations.
``` 
etc...

However there are no areas/groups returned in the next step:
```
--> Model Vocabulary: Cloud Information Model (CIM)
    --> Subject Areas for the Cloud Information Model (CIM) model
    --> Property Groups for the Cloud Information Model (CIM)
```
And hence searched fail:
```
 
Results found for search string: .*Party.*
-------------------------------------------------------------------------------------------
    no terms found
 
 
Terms exactly matching: Party
-------------------------------------------------------------------------------------------
    no terms found
```

to investigate....

**Comment by planetf1** [comment]
In addition:
Still to do (maybe in new PR)
 - ensure tools are easily runnable
 - update docs (including notebook) to clarify how to run/refer to pre-existing archive

**Comment by mandy-chessell** [comment]
I agree that ideally we should ship the CloudInformationModel.json.  However we do not have permission for it.  I have asked repeatedly on the CIM calls if we can do it but no-one knows if it is ok or who should be able to give permission.  I'll escalate through LF.

**Comment by planetf1** [comment]
I wasn't aware of that. Ok so we shouldn't merge the proposed change - at least. I can leave it open briefly, but if the response is negative we'll have to close - or at least update with any new plan.

There's a variety of definitions around 'shipped'. It's often a contentious term. 

In the change I made, the file will appear in
 - An archive that is generated when someone builds the code
 - Forms part of one of our docker images for our demo environment (coco)

It does not
 - appear in a downloaded archive from github
 - appear in any maven artifacts

We could:
* Add an additional *nix container to the environment which we can direct users to, where they can find the utilities and source files (if we're permitted to ship those) and we can direct them to run the command themselves -- more awkward of course
* Add a note to the notebook explaining it will not work in the containerized environments, as well as elaborating on the instructions to build the archive (if permitted)
* Consider whether to make the utilities easier to use - as above, by including dependencies, and optionally having multiple jars with main (allows 'java -jar xxx')

**Comment by ruanwz** [comment]
I was unable to finished the lab because I can't find this CloudInformationModel.json. Could some one tell me where to download this file? Thanks!

**Comment by planetf1** [comment]
This issue is still open -- not yet fixed.

This file is not shipped, nor currently built. There was a prototype fix in https://github.com/planetf1/egeria/tree/issue2278 which you're welcome to view, but we have a few issues to work through including checking if the parser is still working & ensuring licensing is appropriate to package up the generated file.

Basically in open-metadata-resources/open-metadata-archives/design-model-archives you need to run 'org.odpi.openmetadata.archiveutilities.designmodels.cloudinformationmodel.CloudInformationModelArchiveWriter' with a parm of 'CloudInformationModel.json' . This should be doable from within an IDE such as IntelliJ as the classpath is built from the enclosing maven module. 

The work-in-progress above was to do this as part of the build process

Note that as an aside when trying this again I noticed:
```
08:56:37,707 [INFO]      [java] 08:56:37.543 [main] DEBUG org.odpi.openmetadata.repositoryservices.archiveutilities.OMRSArchiveBuilder - Retrieving EntityDef: Referenceable
```

The standard build process has log level set to INFO - reasonable... but these debug messages (2000 of them) are being produced . The code uses debug so appears to be an issue with the logger configuration in the execution of the tool (most likely the way I built the utility in my patch) - will revisit when we address the other issues.

**Comment by planetf1** [comment]
One of the following needs to happen to allow this tutorial to work

a) We add the cloud information model source checked in, build the archive during our build process, and deploy on the egeria server
b) We retrieve the cloud information model via URL during the build process, build the archive, and deploy on the egeria server
c) We provide client tools to build the archive within the containerised environment, guide the user to download manually, and demonstrate how to deploy the model to the server before returning to the notebook

There are seperate issues open to improve our clients and tooling generally.

As such removing milestone from 1.8 and over to @mandy-chessell for ownership until we have something to work with

**Comment by planetf1** [comment]
#2693 adds the utility in a easily usable form into the build and assembly - including then onwards to our docker image & k8s environment.

This change can subsequently look at adding the actual model into the build.

**Comment by planetf1** [comment]
This depends on #3231 to ensure the assemblies are complete

**Comment by planetf1** [comment]
I'm closing #3203 as a dup of this issue since the main issue is that the standard model labs doesn't work (and never has in the container environment). cc: @geoHeil
